DM-34616: prepare for NOT NULL constraints on implied dimensions

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

TallJimbo committed:
Allow "view" dimensions to be passed in test utility code.

This allows "band" data IDs to be passed to MakeTestRepo.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -107,7 +107,8 @@ def makeTestRepo(
     butler = Butler(newConfig, writeable=True)
     dimensionRecords = _makeRecords(dataIds, butler.registry.dimensions)
     for dimension, records in dimensionRecords.items():
-        butler.registry.insertDimensionData(dimension, *records)
+        if butler.registry.dimensions[dimension].viewOf is None:
+            butler.registry.insertDimensionData(dimension, *records)
```
Can you explain why this change is needed? I thought I'd already included support for registering bands.

TallJimbo replied:
It may be that it was already possible via some other method, but I made this change because passing band key in the `dataIds` argument did result in an attempt to insert it, which failed.  I suspect that I missed a way to make it unnecessary to pass a band key here, even after the NOT NULL constraint lands - and I'll discuss that in a thread for another comment - but even if that's the case it seems like it might still be good to permit a `band` key to be passed here, and allow its values to be associated with other dimensions like `physical_filter`, in the same way other relationships are generated arbitrarily.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -183,7 +184,7 @@ def _makeRecords(dataIds: Mapping[str, Iterable], universe: DimensionUniverse) -
             # irrelevant dimensions
             for other in dimension.implied:
-                if other != dimension and other.name in expandedIds and other.viewOf is None:
+                if other.name in expandedIds:
```
If you're doing what I think you're doing, then the comment is no longer correct.

kfindeisen replied:
Part of the reason `_testRepo.py` exists in the first place is to allow `PipelineTask` developers to set up a test environment without any deep knowledge of how Middleware works, including the sometimes unintuitive relationships between dimensions and the special treatment of `band`.

Would it be possible to generate a value (e.g., `band = physical_filter`) analogously to what `_fillAllKeys` does, instead of forcing the user to figure out why their test setup is giving cryptic errors both when they try to create a band and when they don't?

kfindeisen replied:
Do similar changes need to be made to `addDataIdValue`? It's a bit surprising that all the code changes are in the old system.

TallJimbo replied:
I agree the comment is incorrect, and I'll at least fix that (if I can't make it remain correct).

I think there may indeed be an opportunity to generate values for the missing dimensions automatically, and avoid the other branches for this ticket, and I just didn't understand where to put it.  My guess would be to have `_makeRecords` start by ensuring that its input mapping has keys for all dimensions implied by any of the dimensions it is given, using that `_fillAllKeys`-like logic.  I'll attempt that later today, but if you can confirm or reject that as a reasonable approach first I'd appreciate it.

> Do similar changes need to be made to `addDataIdValue`? It's a bit surprising that all the code changes are in the old system.

If I can fix the other code path as you've requested, then I think that means I should fix this one, too.  I suspect the reason I didn't have to was because it already made it possible (but after the NOT NULL constraint lands, necessary) to pass implied values by including them in the data ID key, so I just went around making changes in the tests like:
```
        -addDataIdValue(cls.creatorButler, "physical_filter", "l2019", instrument="dummyCam")
        +addDataIdValue(cls.creatorButler, "physical_filter", "l2019", band="l", instrument="dummyCam")
```
And I think the right place for that fix is in `_matchAnyDataId`, which could create (with `_findAllKeys`-like logic) and insert a new dimension record when there isn't one, instead of raising.  Once again, confirmation that that approach makes sense would be appreciated.

kfindeisen replied:
>  My guess would be to have `_makeRecords` start by ensuring that its input mapping has keys for all dimensions implied by any of the dimensions it is given, using that `_fillAllKeys`-like logic. I'll attempt that later today, but if you can confirm or reject that as a reasonable approach first I'd appreciate it.

Yes, I think so. Unfortunately, I haven't looked at this code since last year, and I wrote it with a very vague idea of how the dimensions system works, so I can't offer any special insights.

> And I think the right place for that fix is in `_matchAnyDataId`, which could create (with `_findAllKeys`-like logic) and insert a new dimension record when there isn't one, instead of raising. Once again, confirmation that that approach makes sense would be appreciated.

I'm a little nervous about having `addDataIdValue` create dimension records for other dimensions (looks like it would be `physical_filter` in practice), but I think it's the least of the available evils. Please add a note to the docstring that it may create values for dimensions other than `dimension`.

kfindeisen commented on tests/test_testRepo.py:
```diff
@@ -53,7 +53,7 @@ def setUpClass(cls):
         addDataIdValue(cls.creatorButler, "instrument", "dummyCam")
         addDataIdValue(cls.creatorButler, "physical_filter", "k2020", band="k", instrument="notACam")
-        addDataIdValue(cls.creatorButler, "physical_filter", "l2019", instrument="dummyCam")
+        addDataIdValue(cls.creatorButler, "physical_filter", "l2019", band="l", instrument="dummyCam")
```
I think this line is actually redundant now, since its purpose was to test that you could omit `band`. Does the test suite break if you delete this line?

TallJimbo replied:
Yes, and it sounds like the preferred approach is to revert this change and try to make the test pass without it (as per the other thread).

kfindeisen replied:
If the test breaks, then the line is needed either way.

kfindeisen reviewed (commented):
I think I understand the problem, but this feels more like a workaround than a solution. The test support code previously *made up* values as necessary to get valid data IDs and dimension records; why can't the same approach work for implied dimensions?

kfindeisen commented:
One other comment: if the system can't be improved over what's here, then https://pipelines.lsst.io/v/daily/modules/lsst.daf.butler/use-in-tests.html#creating-and-populating-a-repository probably needs to be updated with a sentence saying that you need to define at least one value for most dimensions (and the examples need vetting to see if they're still valid).

TallJimbo committed:
Automatically generate implied dimension records in test utility code.

This required a fair amount of restructuring of the test utility code
implementation.  Some variable names were changed as well to better
reflect nomenclature used elsewhere.

This required some changes to unit tests because it's not possible to
distinguish between cases where the caller wants some dimension records
created automatically and cases where the caller defined a data ID
that is inconsistent with existing records - the logic now just creates
records as necessary to make the given data ID valid.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -394,6 +395,11 @@ def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
+    if fullDimension.viewOf:
+        # Nothing to do; this dimension's records aren't actually saved in the
+        # database directly anyway (and caller shouldn't have to know this).
+        return
```
Clear and concise! 👍

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -57,11 +57,12 @@ def makeTestRepo(
+        A mapping keyed by the dimensions used in the test. Each value is an
+        iterable of names for that dimension (e.g., detector IDs for
+        `"detector"`). Related dimensions (e.g., instruments and detectors) are
+        linked arbitrarily, with values created randomly when needed. This
```
I would omit "randomly" from this statement; I would prefer that the user make no assumptions whatsoever about what the values do or don't look like.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -165,36 +166,41 @@ def _makeRecords(dataIds: Mapping[str, Iterable], universe: DimensionUniverse) -
+        if dimension.name in dataIds:
+            complete_data_id_values[dimension.name] = list(dataIds[dimension.name])
+        if dimension.name not in complete_data_id_values:
+            complete_data_id_values[dimension.name] = [_makeRandomDataIdValue(dimension)]
```
I find it very strange that Codecov is claiming that neither of these blocks is entered in tests. That shouldn't be the case if there's even a single test with non-empty `dataIds`.

TallJimbo replied:
It looks like there just are no tests of `dataIds` in daf_butler; the ones here all use `addDataIdValue` instead, which fits much better with the structure of the test, since it tries to use only one repo created in `setUpClass` (for performance reasons).  There are tests in downstream packages that use `dataIds`.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -165,36 +166,41 @@ def _makeRecords(dataIds: Mapping[str, Iterable], universe: DimensionUniverse) -
 
 
-def _fillAllKeys(dimension: Dimension, value: Any) -> Mapping[str, Any]:
+def _fillAllKeys(dimension: Dimension, value: Union[str, int]) -> Mapping[str, Union[str, int]]:
```
Can you actually guarantee that all dimensions (including any added in the future) will be either strings or integers?

timj replied:
All current ones are. Floating point dimensions are problematic and we've had discussion about what to do if we have spectrographs that are tunable -- what's the equivalent of physical filter for a heterodyne receiver that can tune to arbitrary frequency and have configurable bandwidth? Should the central tuning frequency be used in the dataId? If it's 230.23 GHz should it be converted to an integer of 230230 for that instrument? Does every possible floating point value have to have a registered dimension record or can it be more flexible?

TallJimbo replied:
Yes, we can guarantee that - or at least it's an assumption also made elsewhere in daf_butler, and encoded in other type annotations like this one.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -215,7 +221,11 @@ def _fillAllKeys(dimension: Dimension, value: Any) -> Mapping[str, Any]:
+            # except clause below, and that in particular is a problem because
+            # of the skymap dimensions' bytes 'hash' field, which has a
+            # unique constraint.
+            castType = lambda x: str(x).encode()  # noqa: E731
```
Can you explain why this is needed? Making `castType` be anything other than a type makes the subsequent code somewhat misleading; I'm not sure if the "default-construction" fallback still works, either.

TallJimbo replied:
Without this, the `_findAllKeys` logic for `bytes` fields always falls through to the `except` clause below, because `bytes(int)` `bytes(str)` both raise, and hence the value is always just `b''`.  When there is a unique constraint on a bytes field - as there is for the `skymap` dimension, and when more than one record in inserted (which is more likely now that dimension values are sometimes invented as needed, and one of the tests now does it), we get a constraint violation.

I think you're right that the fallback no longer works with this, though; I'll see about fixing that.

kfindeisen replied:
Got it, so it's a bug in my code. However, can you reword the comment to explicitly say that this is, logically, a conversion to `bytes`? That won't be clear once the diff is gone.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -165,36 +166,41 @@ def _makeRecords(dataIds: Mapping[str, Iterable], universe: DimensionUniverse) -
-                    relation = expandedIds[other.name][0]
-                    value[other.name] = relation[other.primaryKey.name]
+                    relation = record_dicts_by_dimension_name[other.name][0]
+                    record_dict[other.name] = relation[other.primaryKey.name]
```
I think I'm a bit confused about the scope of the change. The old code made a distinction between required dimensions (assumed to be provided by the user) and implied dimensions (hidden from the user). The new code tries to hide everything. Is there a reason continuing to require required dimensions is now impractical?

TallJimbo replied:
A version that invents required dimensions as needed as well was a bit easier to implement, and I figured it'd actually be desirable behavior if the overall goal is keeping tests from caring about the details of the dimension data model.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -350,14 +311,14 @@ def expandUniqueId(butler: Butler, partialId: Mapping[str, Any]) -> DataCoordina
 
 
-def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
+def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **related: Union[str, int]):
```
Same question about how safe it is to assume it's only these two types.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -350,14 +311,14 @@ def expandUniqueId(butler: Butler, partialId: Mapping[str, Any]) -> DataCoordina
+    specified using ``related``. While these keywords are sometimes needed to
+    get self-consistent repositories, you do not need to define relationships
+    you do not use. Any unspecified dimensions will be linked arbitrarily,
+    withe values created randomly when needed.
```
Typo, "withe". I'd prefer not to say *how* the values are created.

kfindeisen replied:
I'm worried the existing text may not make the consequences clear enough. Can the "Notes" section say something like "values for dimensions other than `dimension` may be created if necessary to create a valid instance of `dimension`".

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -388,32 +349,71 @@ def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
         raise ValueError(
-            f"Unexpected keywords {extraKeys} not found "
-            f"in {fullDimension.required | fullDimension.implied}"
+            f"Unexpected keywords {extraKeys} not found " f"in {fullDimension.graph.dimensions.names}"
```
Weird formatting -- please make this either a single string, or on two lines.

timj replied:
That's black realizing that it can now fit it in one line, but it does not attempt to merge the strings. It's a black oddity that will happen and we need to keep an eye out for.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -388,32 +349,71 @@ def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
+    # This iteration is guaranteed to be in topological order, so we can count
+    # on new data ID values being invented before they are needed.
+    record_dicts_by_dimension: dict[Dimension, dict[str, Any]] = {}
+    for dimension_obj in all_dimensions:
```
I'm finding this loop very hard to follow because of the lack of distinction between the dimension we are actually trying to create (`data_id`) and the dimensions that are incidental to that and probably already exist (`dimension_obj`, `dimension_value`, etc.). I especially don't understand why you are carefully constructing dimension records for existing dimensions.

Would it be possible to factor out some of these blocks into their own functions (making the code a bit more self-documenting in the process)? It looks like the body of `if dimension_value is None` and the code below the view check are each trying to do one thing. This should reduce the number of variables the human needs to keep track of at a time.

TallJimbo replied:
In a system where any missing dependencies are created as needed, I don't think there is a distinction between the original dimension identified and any others anymore; it's unfortunate that the interface continues to suggest that there is, but I don't want to go through the deprecation-cycle effort of changing that now, either.

>  especially don't understand why you are carefully constructing dimension records for existing dimensions.

I don't think I am; I am fetching any records that exist that are compatible with the given data IDs.  Calling `syncDimensionData` on those later should be a no-op (it's an expensive enough one that I would try to avoid it in non-test code, but here I figured that was better than having another local variable to track which records are new and which are existing).

> Would it be possible to factor out some of these blocks into their own functions

:+1:

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -388,32 +349,71 @@ def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
+            if related_dimension.name != dimension_obj.name:
+                record_dict[related_dimension.name] = data_id[related_dimension.name]
+
+        record_dicts_by_dimension[dimension_obj] = record_dict
```
Would it be simpler to just call `addDataIdValue` recursively for missing required dimensions? I'm worried about `syncDimensionData` either failing, or worse, changing, records for dimensions that existed before the call.

TallJimbo replied:
`syncDimensionData` definitely will not change existing records, and I think if it does fail it does still say something important.  That said, I'm not sure I've left it possible for it to fail - I have not been able to trigger the previous expected failures in the tests without breaking something else (more on that in other threads).

TallJimbo replied:
I did look a bit into calling `addDataIdValue` recursively just now, and I think it'd a bit messier in the end, because we'd have to add new arguments to distinguish between the original contents of `**related` and later derived ones.

kfindeisen commented on tests/test_testRepo.py:
```diff
@@ -126,15 +126,16 @@ def testAddDataIdValue(self):
 
-        # Keywords imply different instruments
-        with self.assertRaises(RuntimeError):
-            addDataIdValue(self.butler, "exposure", 101, instrument="dummyCam", physical_filter="k2020")
```
I don't understand why this test case has been removed. The extra dimensions are inconsistent (`"k2020"` implies `"notACam"`, not `"dummyCam"`), so it should still fail.

TallJimbo replied:
I played around with a number of adjustments to try to make this work, and eventually I convinced myself that if the system is allowed to create new dimension values, there's no way to distinguish between this case and missing implied dimension records more generally, at least not without the kind of exhaustive search that the butler deliberately disables because in real-world contexts it always indicates a mistake (if someone queries for data IDs with `detector=4` in a real repo without specifying the instrument, they're not asking for to cartesian-product iterate over all instruments that have a detector 4).

Switching to only allowing implied dimension records to be created might mitigate this somewhat, but it wouldn't save this test (that's one of the things I tried), because `physical_filter` is an implied dimension of `exposure`, so creating one that matches the undesired `instrument` with the given filter name would still be permitted.

So I think the choice is between this behavior and reverting back to the original version of this branch, where downstream tests were still required to provide implied dimension value explicitly.

TallJimbo replied:
Ah!  Nevermind - the thread just below this one made me realize the distinction you want made: you do not want records created for any dimension that is referenced explicitly in the call to `addDataIdValue`; those are asserted to already exist, and only those not mentioned should be created automatically.  It seems obvious now; I was too deep in the weeds (having basically dropped the original set of dimensions given on the floor in favor of its expansion) to see it before.  I'll be back with another of this shortly.

TallJimbo replied:
Ok, done!  This let me restore the original test to exactly what it was before.

kfindeisen commented on tests/test_testRepo.py:
```diff
@@ -126,15 +126,16 @@ def testAddDataIdValue(self):
+        self.assertEqual(tract_data_id["tract"], 42)
+        # And then a new one created here, since the now-existing one should
+        # have a random name that isn't 'map'.
+        addDataIdValue(self.butler, "tract", 43, skymap="map")
```
Please do not allow this behavior; it will be too surprising for users. A user who specified a specific skymap is clearly assuming it exists.

TallJimbo replied:
I think this is a case that would be addressed by permitting only implied dimension records to be created.

kfindeisen reviewed (changes requested):
Apologies for dragging this out, but while the code itself looks reasonable (if hard to understand), the test cases imply behavior that sharply deviates from [what's been documented](https://pipelines.lsst.io/v/daily/modules/lsst.daf.butler/use-in-tests.html#creating-and-populating-a-repository).

I'm particularly concerned about new values being created for dimensions where the user was almost certainly expecting to use existing ones and/or overwriting dimensions the user previously created. The behavioral changes turn the purpose of `addDataIdValue` from "create one dimension, and sweep the implementation details under the rug" to "create multiple dimensions anchored at a particular node".

(Also, in the future, please put formatting and variable renames on different commits from changes of substance. The blend made it very hard for me to reconstruct how the new code did or did not relate to the old code.)

TallJimbo reviewed (commented):
Ok, I've just pushed a change to address the last round of comments (will squash later), and I'm pretty happy with it.  I also hit "resolved" on a few threads here where the resolution was obvious.

The big good news is that the original test now passes, so this is really the minimal change needed to address the upcoming problem.

The code coverage is apparently not great:
- the existing coverage wasn't as good as either of us expected, though there is coverage in downstream packages, and what's missing here is clearly documented as the "old interface";
- the new logic that adds implied dimension records isn't tested here _yet_, but it will be covered by the existing tests as soon as the new constraint lands.

I think that's enough for me, even if it's not ideal.

Finally, apologies for mixing up the renames with the behavior changes - separating those is usually very much my policy as well, but in this case I ended up doing the renames as I worked out what the code was doing, and by the time I was done I figured the functions I touched were pretty much completely rewritten so the quality of the diff wasn't as important (though it seems the PR diff pretty thoroughly obfuscated that).

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -60,9 +60,9 @@ def makeTestRepo(
-        linked arbitrarily, with values created randomly when needed. This
-        parameter is provided for compatibility with old code; newer code
-        should make the repository, then call
+        linked arbitrarily, with values created for implie dimensions only when
```
```suggestion
        linked arbitrarily, with values created for implied dimensions only when
```

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -221,11 +221,13 @@ def _fillAllKeys(dimension: Dimension, value: Union[str, int]) -> Mapping[str, U
+            # skymap dimensions' bytes 'hash' field, which has a unique
+            # constraint.
+            # This lambda casts `str` and `int` values to bytes and yields b''
+            # when called with no arguments (as in the except block below).
```
I think it would make more sense to say this first, and *then* justify why you're using a lambda to do it.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -311,14 +313,88 @@ def expandUniqueId(butler: Butler, partialId: Mapping[str, Any]) -> DataCoordina
     get self-consistent repositories, you do not need to define relationships
-    you do not use. Any unspecified dimensions will be linked arbitrarily,
-    withe values created randomly when needed.
+    you do not use.  Any unspecified dimensions implied will be linked
```
"Any unspecified implied dimensions"? I'm having trouble parsing the sentence as-is.

TallJimbo replied:
I've rewritten this block and merged it into to the `Notes` section, as it was getting pretty long.  Here's the new complete `Notes` text (the second paragraph is unchanged):

> Related dimensions (e.g., the instrument associated with a detector) may be
    specified using ``related``, which requires a value for those dimensions to
    have been added to the repository already (generally with a previous call
    to `addDataIdValue`.  Any dependencies of the given dimension that are not
    included in ``related`` will be linked to existing values arbitrarily, and
    (for implied dependencies only) created and also inserted into the registry
    if they do not exist.  Values for required dimensions and those given in
    ``related`` are never created.
>
>    Because this function creates filler data, it is only suitable for test
    repositories. It should not be used for repositories intended for real data
    processing or analysis, which have known dimension values.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -337,6 +413,13 @@ def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **rel
+    in order to construct records for ``dimension`` that satisfy database
+    foreign key constraints.  This is never done for required dimensions (those
+    that are also needed disambiguate any value for ``dimension``) or those
+    given as keys in ``related``; records fo hose dimensions must already
```
```suggestion
    given as keys in ``related``; records for those dimensions must already
```

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -311,14 +313,88 @@ def expandUniqueId(butler: Butler, partialId: Mapping[str, Any]) -> DataCoordina
+        return dimension_value, True
+    else:
+        # A record does exist in the registry.  Use its data ID value,
+        # and skip the rest of the record-fabrication logic.
```
The "skip" part of this statement is no longer applicable, since it's outside the scope of this function.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -311,14 +313,88 @@ def expandUniqueId(butler: Butler, partialId: Mapping[str, Any]) -> DataCoordina
+    data_id : `dict` [ `str`, `str` or `int` ]
+        Dictionary that contains values for all of
+        ``dimension.dimensions.names`` (the main dimension, its recursive
+        required dependencies, and its non-recursive implied dependencies).
```
Are the implied dependencies non-recursive? I thought satisfying dependencies recursively was the point of this PR.

TallJimbo replied:
The record for a particular dimension only includes direct (non-recursive) implied dimension values (e.g. `visit` has a column for `physical_filter`, but not `band), which is why that's all that's needed in this method.  But the records for those implied dependencies will then include columns for their implied dependencies, so we do need to recurse elsewhere.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -370,45 +453,33 @@ def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **rel
+            # isn't the primary dimension they asked to have a record created
+            # for.  That means they expect this record to already exist.
+            continue
+        if dimension_obj != fullDimension and dimension_obj in all_dimensions.required:
+            # We also don't want to automatically create new dimension records
+            # for required dimensions (except for the main dimension the caller
+            # asked for); those are also asserted by the caller to already
+            # exist.
+            continue
```
This was the original behavior, but I thought (based on your answer to https://github.com/lsst/daf_butler/pull/680#discussion_r871800711) that this wouldn't work anymore. Given the incoming not-null constraints, what "non-obvious" dimensions will people need to specify that they didn't before? `physical_filter` for `exposure`?

TallJimbo replied:
The original behavior was that users did not need to provide values for implied dependencies, and the foreign key columns with the values for those dimensions were set to `None` / `NULL`.

The new behavior is that users still do not need to provide values for implied dependencies, but this is now because those missing values are generated randomly and records for them are inserted into the registry, allowing the foreign key columns that point to them to be not-`NULL` and have their constraints satisfied.

kfindeisen replied:
My mistake, I'd forgotten the meaning of "required" (as opposed to "implied") in the context of dimensions.

kfindeisen commented on python/lsst/daf/butler/tests/_testRepo.py:
```diff
@@ -370,45 +453,33 @@ def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **rel
         try:
             butler.registry.syncDimensionData(dimension_obj, record)
-        except ConflictingDefinitionError as e:
+        except sqlalchemy.exc.IntegrityError as e:
```
I don't understand this change. Was it a mis-specification of `syncDimensionData`?

TallJimbo replied:
Short version: this is reverting the exception type back to `sqlalchemy.exc.IntegrityError`, which it was originally and probably should have been throughout my work on the branch, but in the middle I got confused about what the likely failure mode was.

Long version: `IntegrityError` is the right type for low-level foreign key and unique constraints (in this context, foreign key violations are much more likely, and what the re-raise here anticipates, but sqlalchemy makes it really hard to tell which, which is the main reason we don't re-raise at a lower level).  The changes on this branch that have not been dropped made that kind of constraint violation impossible, but I wasn't able to convince myself that a different error condition could be ruled out, in which the record already exists but it has different content, and `ConflictingDefinitionError` is raised.  So I switched the except block to that.  With the final version of the logic here that's more like the original, foreign key violations (due to missing required dependencies) are once again possible, and the original re-raise makes sense again.

kfindeisen commented on tests/test_testRepo.py:
```diff
@@ -126,16 +126,15 @@ def testAddDataIdValue(self):
+        with self.assertRaises(RuntimeError):
+            addDataIdValue(self.butler, "tract", 42)
+        with self.assertRaises(RuntimeError):
+            addDataIdValue(self.butler, "tract", 43, skymap="map")
```
Maybe this should have its own comment:
```suggestion
        # Didn't create skymap "map" first.
        with self.assertRaises(RuntimeError):
            addDataIdValue(self.butler, "tract", 43, skymap="map")
```

kfindeisen reviewed (approved):
Looks good, mostly documentation comments left. Thank you for your patience!

My biggest outstanding question is about exactly which dependencies do  or don't get invented. I just want to make sure the current solution satisfies both our requirements (that existing tests will not break when the new registry constraints come in, and that pipeline task developers don't need more knowledge of the dimensions system than they needed to write the task in the first place).
